### PR TITLE
Admin user permissions

### PR DESCRIPTION
**Pull request by tsiology**
Commenting out automatic OrgUser object creation from organization signup methods. OrgUser creation is under review while determining how to create to org admins.

### CODE REVIEW

**Comment by nickolashe** [comment]
@tsiology Is this ready for review / testing?

**Comment by tsiology** [comment]
@nickolashe yes, review requested. Simply commenting out OrgUser creations for now

**Comment by nickolashe** [comment]
@tsiology I think I'd like to handle this differently. Lets still record the orguser association, but lets also add an approved field to that table. Anywhere we are checking for an orguser association, we'd also make that conditional on this new field.

**Comment by nickolashe** [comment]
This will allow us to manually review these orguser association requests.

**Comment by nickolashe** [comment]
Note that orguser also an affiliation field, so we should also be checking whether that's equal to `admin` whenever deciding if the user has admin permissions.

**Comment by tsiology** [comment]
@nickolashe I've added back the OrgUser creations but with a new field approved defaulted to False included in the table. I'll need to review all instances of where we filter OrgUser objects and all the check for approved=True. For affiliation=admin, are we ready to start filtering based off that as well?

**Comment by nickolashe** [comment]
@tsiology yep, i think we are

**Comment by nickolashe** [comment]
@tsiology let's also start using django's permissions for the purpose of granting access. please see my last commit.

**Comment by nickolashe** [comment]
as well as this section from the docs:
https://docs.djangoproject.com/en/1.11/topics/auth/default/#permissions-and-authorization

**Comment by nickolashe** [comment]
i.e., once we approve an admin user (manually, for now), they will be granted the `admin` permission

**Comment by dannypernik** [comment]
@nickolashe so an OrgUser will not have an orgid until manually approved? We have {% if orgid %} logic in several places so want to verify an unverified OrgUser will not see admin features.

**Comment by nickolashe** [comment]
@dannypernik did our recent convo help resolve your question above?

**Comment by nickolashe** [comment]
#155

**Comment by nickolashe** [comment]
#213

File: openCurrents/views.py [hunk 1]
```diff
@@ -619,6 +619,33 @@ def get_context_data(self, **kwargs):
...
+        #check that user is part of org_admins group
+        userid = self.request.user.id
+        user = User.objects.get(id=userid)
+        org_admins_name = 'admin_' + str(orgid)
+        logger.info('org_admins_name: %s',org_admins_name)
+
+        try:
+            org_admins = Group.objects.get(name=org_admins_name)
+        except:
+            logger.info("Org exists without an org_admins group.")
```
Comment (nickolashe) [inline openCurrents/views.py]: lets log this at error level

File: openCurrents/views.py [hunk 2]
```diff
@@ -2017,6 +2048,12 @@ def process_org_signup(request):
 
         try:
             org.save()
+
+            org_admins_name = 'admin_' + str(org.id)
```
Comment (nickolashe) [inline openCurrents/views.py]: I think the fact that this block of code is repeating the one above warrants factoring it out into a function

File: openCurrents/views.py [hunk 3]
```diff
@@ -619,6 +619,33 @@ def get_context_data(self, **kwargs):
...
+
+        try:
+            org_admins = Group.objects.get(name=org_admins_name)
+        except:
+            logger.info("Org exists without an org_admins group.")
+            pass
+
+        if org_admins and not user.groups.filter(name=org_admins_name).exists():
+            logger.info("org admins group exists and user is not part of it.")
+            return redirect('openCurrents:500')
```
Comment (nickolashe) [inline openCurrents/views.py]: the error should be 403 (forbidden)

File: openCurrents/views.py [hunk 4]
```diff
@@ -619,6 +619,33 @@ def get_context_data(self, **kwargs):
...
+        userid = self.request.user.id
+        user = User.objects.get(id=userid)
+        org_admins_name = 'admin_' + str(orgid)
+        logger.info('org_admins_name: %s',org_admins_name)
+
+        try:
+            org_admins = Group.objects.get(name=org_admins_name)
+        except:
+            logger.info("Org exists without an org_admins group.")
+            pass
```
Comment (nickolashe) [inline openCurrents/views.py]: this needs to return an http response of some kind, probably a 500 error page

File: openCurrents/views.py [hunk 5]
```diff
@@ -619,6 +619,33 @@ def get_context_data(self, **kwargs):
 class AdminProfileView(LoginRequiredMixin, SessionContextView, TemplateView):
     template_name = 'admin-profile.html'
 
+    def get(self, request, *args, **kwargs):
+        context = super(AdminProfileView, self).get_context_data(**kwargs)
+        orgid = context['orgid']
```
Comment (nickolashe) [inline openCurrents/views.py]: i think we can improve on our current assumption of a single user to org mapping and check all such mappings for the purpose of authorizing the user

Comment (tsiology) [inline openCurrents/views.py]: @nickolashe can you expand on this? I think you are saying context['orgid'] should be a list of all orgid's the user/admin is associated with and we should check all org_admin groups for those ids.

**Comment by tsiology** [comment]
@nickolashe I've removed the overridden get in AdminProfileView approach to instead set a context variable forbidden in get_context_data. This variable is then checked in the template. Functionality of redirecting users who are not in the group from admin-profile to 500 is working. Using 500 till 403 page is available.

File: openCurrents/templates/admin-profile.html [hunk 6]
```diff
@@ -255,6 +254,14 @@ <h3 class="popup-title">Volunteers request approval</h3>
       });
     </script>
 
+
+    <script type="text/javascript">
```
Comment (nickolashe) [inline openCurrents/templates/admin-profile.html]: @tsiology what if the user turns off javascript in their client?

Comment (nickolashe) [inline openCurrents/templates/admin-profile.html]: i think we want to still stay within django realm

Comment (nickolashe) [inline openCurrents/templates/admin-profile.html]: @tsiology see if you can override `render_to_response` method, which executes after `get_context_data` for class-based views:
https://docs.djangoproject.com/en/1.11/ref/class-based-views/mixins-simple/#django.views.generic.base.TemplateResponseMixin.render_to_response

Comment (nickolashe) [inline openCurrents/templates/admin-profile.html]: also, FYI here is the list of class-based view methods listed in their execution order:
https://docs.djangoproject.com/en/1.11/ref/class-based-views/generic-display/#detailview

Comment (tsiology) [inline openCurrents/templates/admin-profile.html]: would this override of render_to_response take the place of render in the get function as a way to pass all needed context to the template?

Comment (nickolashe) [inline openCurrents/templates/admin-profile.html]: yes, i think it'll take place of the get. it returns the class instance of the view from what i understand. so no raw rendering like with render

**Comment by nickolashe** [comment]
perhaps lets find an example of overriden render_to_response

File: openCurrents/views.py [hunk 7]
```diff
@@ -88,9 +90,50 @@ def get_context_data(self, **kwargs):
         if userorgs:
             org = userorgs[0].org
             context['orgid'] = org.id
+            context['org_id'] = org.id
 
         return context
 
+class OrgAdminPermissionMixin(LoginRequiredMixin):
```
Comment (nickolashe) [inline openCurrents/views.py]: @tsiology @karbmk I put together this new mixin which can be used to assert admin permissions for any view

File: openCurrents/views.py [hunk 8]
```diff
@@ -88,9 +90,50 @@ def get_context_data(self, **kwargs):
...
+    def dispatch(self, request, *args, **kwargs):
```
Comment (nickolashe) [inline openCurrents/views.py]: It overrides the default `dispatch` method for the view

**Pull request merged**